Sources/ComposableArchitecture/Effects/Cancellation.swift

```diff
@@ -27,45 +56,55 @@ extension Effect {
   ///     canceled before starting this new one.
   /// - Returns: A new effect that is capable of being canceled by an identifier.
   public func cancellable(id: AnyHashable, cancelInFlight: Bool = false) -> Effect {
-    return Deferred { () -> Publishers.HandleEvents<PassthroughSubject<Output, Failure>> in
-      let subject = PassthroughSubject<Output, Failure>()
-      let uuid = UUID()
+    return .deferred { () -> SignalProducer<Value, Error> in
+      let subject = Signal<Value, Error>.pipe()      
 
-      var isCleaningUp = false
+      var disposable: Disposable?
+      var values: [Value] = []
+      var isCaching = true
 
       cancellablesLock.sync {
         if cancelInFlight {
-          cancellationCancellables[id]?.forEach { _, cancellable in cancellable.cancel() }
+          cancellationCancellables[id]?.forEach { cancellable in cancellable.dispose() }
           cancellationCancellables[id] = nil
         }
 
-        let cancellable = self.subscribe(subject)
-
-        cancellationCancellables[id] = cancellationCancellables[id] ?? [:]
-        cancellationCancellables[id]?[uuid] = AnyCancellable {
-          cancellable.cancel()
-          if !isCleaningUp {
-            subject.send(completion: .finished)
+        disposable = self
+          .on {
+            guard isCaching else { return }
+            values.append($0)
```

Comment:
Slightly dubious about this caching approach, but I couldn't see any alternative solution 😕 This is because in Combine `subscribe(subject)` doesn't immediately start publisher, whereas ReactiveSwift's `start(subject.input)` does, so any values that are sent immediately are lost, unless we do this caching.